.buildkite/pipeline_generator/pipeline_generator.py

```diff
@@ -0,0 +1,180 @@
+import yaml
+import click
+from typing import List, Dict, Union
+import os
+
+from plugin import (
+    get_kubernetes_plugin_config,
+    get_docker_plugin_config,
+)
+from utils import (
+    AgentQueue,
+    AMD_REPO,
+    TEST_PATH,
+    EXTERNAL_HARDWARE_TEST_PATH,
+    PIPELINE_FILE_PATH,
+    STEPS_TO_BLOCK,
+    get_agent_queue,
+    get_full_test_command,
+    get_image_path,
+    get_multi_node_test_command,
+)
+from step import (
+    TestStep, 
+    BuildkiteStep, 
+    BuildkiteBlockStep, 
+    get_block_step, 
+    get_step_key
+)
+
+
+def _get_plugin_config(step: TestStep) -> Dict:
+    """
+    Returns the plugin configuration for the step.
+    If the step is run on A100 GPU, use k8s plugin since A100 node is on k8s.
+    Otherwise, use Docker plugin.
+    """
+    test_bash_command = [
+        "bash", 
+        "-c", 
+        get_full_test_command(step.commands, step.working_dir)
+    ]
+    docker_image_path = get_image_path()
+
+    if step.gpu == "a100":
+        return get_kubernetes_plugin_config(
+            docker_image_path, 
+            test_bash_command
+        )
+    return get_docker_plugin_config(
+        docker_image_path, 
+        test_bash_command, 
+        step.no_gpu
+    )
+
+def read_test_steps(file_path: str) -> List[TestStep]:
+    """Read test steps from test pipeline yaml and parse them into Step objects."""
+    with open(file_path, "r") as f:
+        content = yaml.safe_load(f)
+    return [TestStep(**step) for step in content["steps"]]
+
+def process_step(step: TestStep, run_all: str, list_file_diff: List[str]) -> List[Union[BuildkiteStep, BuildkiteBlockStep]]:
+    """Process test step and return corresponding BuildkiteStep."""
+    steps = []
+    current_step = BuildkiteStep(label=step.label, key=get_step_key(step.label), parallelism=step.parallelism, soft_fail=step.soft_fail, plugins=[_get_plugin_config(step)])
+    current_step.agents["queue"] = get_agent_queue(step.no_gpu, step.gpu, step.num_gpus).value
+
+    # If the step is a multi-node test, run it with a different script instead of Docker plugin
+    if step.num_nodes > 1:
+        current_step.commands = get_multi_node_test_command(
+            step.commands, 
+            step.working_dir, 
+            step.num_nodes, 
+            step.num_gpus, 
+            get_image_path()
+        )
+        current_step.plugins = None
+
+    def step_should_run():
+        """Determine whether the step should automatically run or not."""
+        if not step.source_file_dependencies or run_all == "1" or step.gpu == "a100":
+            return True
+        return any(source_file in diff_file 
+            for source_file in step.source_file_dependencies 
+            for diff_file in list_file_diff)
```

Comment:
why does this need to be an inlined function?

Reply:
cause it reuses the `step` object ... I guess I can also separate it as another method ..